### ISSUE TITLE
Directory cleaner

### ISSUE THREAD

**Issue by aurrelhebert** [senx/warp10-platform#1107]
# Directory cleaner

Hello Warp10 community, we are a big user of the Warp10 platform, and might be the first one to reach a billion series. However, this is way too much (at least for our use case). This is not something due to the technology in itself, but mainly how we use it. 

## Ephemeral series

When we write our series data model, we choose to create ephemeral series. At first, it seems to be the best solution as it would avoid having our metrics overlapping.  However soon, the drawback of this choice was that we generated way too much series, which quickly became obsolete. This means that our directories took more and more ram, they also were slower to load their metadata from HBase. But what is now worse is: that a lot of the series have the same class slowing all directories requests. This means that we cannot easily delete inactive series by executing a WarpScript.

## Solution

We needed an efficient way to delete our series without generating too much traffic on our platform. We decided to boot a new server which will host a custom directory a little bit modified. We thought about something similar to the HDFS balancer that we can control with some `systemd timers` or `cron`. 

This new directory will read HBase, and based on a given parameter clean the inactive series. We compare the current date and the last activity of each series we read from HBase. If the inactivity window is above the parameter set in the `directory cleaner` configuration, we send a delete message in Kafka. This means that he has its own producer, and that all other directories up will be notified of the deletion. The `directory cleaner` can then be booted on a new server and is fully independent. In our case, we wanted that when it finishes its jobs to stop. This enables us to boot this new `directory cleaner` with a lot less RAM than our directories. 

## Customisable

In our code, we add a lot of new parameters to set this new feature as customisable as generic as possible. 

We added a boolean parameter whether the directory should stop at the end of the process. 
We added a global time window (ex: 2 years) to delete series that are inactive (ex: series inactive for 2 years).
We added a pair that can contain a map of applications and a time window. This parameter overwrites the previous check to enable series deletion series with a custom-time window for selected applications.
We added the same parameters as an ingress node to customise the directory Kafka producer.

## In conclusion

We have just started to use it on our cluster and started to clear our directories. Now that we start to use it, we want to share it with the community as it can help other people that have the same experience with ephemeral series (or only some Warp10 applications that would have ephemeral series).

We will appreciate any feedback, and/or how could we improve such a component?

**Comment by hbs** [comment]
You mention "This means that we cannot easily delete inactive series by executing a WarpScript.", could such deletions be performed using a call to the /api/v0/delete endpoint, possibly with the use of the 'metaonly' flag to simply remove the metadata?

**Comment by hbs** [comment]
On the implementation side of things, what you describe looks a lot like a Directory plugin, why modify the Directory class instead of creating a Directory plugin?

**Comment by aurrelhebert** [comment]
Indeed, it's not possible for us to do a delete with `metalonly` as selected the series to delete with a `find` is too costly

**Comment by aurrelhebert** [comment]
With some second thought, I do not think it matched the `directory plugin` pattern, as I am adding context when initialising the directory with HBase (in case the la is too old send a message to HBase). Otherwise I keep the same pattern, and the directory work as usual. It's only in the case where the parameter `directory.cleaner.stop` is set to true, that I change the behavior (storing no series in RAM and shutting down the directory when HBase was read). This parameter should may be have an other name, as even if I use it for the `directory cleaner` it can be used globally for the directory.

**Comment by StevenLeRoux** [comment]
We discussed with @hbs  and there could be a possibility as a `trusted directory plugin` which require Mathias to merge it :) Once done we will give it a look, but as I understood it, this case would have all necessary context to arbitrate a deletion while keeping the same usage pattern or running a job regularly.

**Comment by pi-r-p** [comment]
Hello @StevenLeRoux  @aurrelhebert ,
2.11.0 is released with [TrustedDirectoryPlugin](https://github.com/senx/warp10-platform/blob/master/warp10/src/main/java/io/warp10/warp/sdk/TrustedDirectoryPlugin.java).  Waiting for your feedback...
